### PR TITLE
Kb2seurat

### PR DESCRIPTION
**Pull request by tilschaef**
**What problem is the PR solving / What's new?**
Addresses the features described in #745.  This PR implements support for automatically exporting kb count tables to Seurat objects, merging multiple objects. and storing the output in RData format. Moreover, meta-data defined in samples.tsv will be automatically assigned to the cell identifiers in the corresponding Seurat object. 

**What did change**
Added quantification rules` export_seurat_obj /merge_seurat_obj` as well as helper R scripts to implement the functionality described above.

**Checklist**
- [x] I made a PR to develop (not master)
- [ ] If applicable: I updated the docs
- [x] I updated the CHANGELOG
- [ ] These changes are covered by the tests

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -44,6 +48,7 @@ All changed fall under either one of these types: `Added`, `Changed`, `Deprecate
   - the samples.tsv will now work with only 2 columns
   - the samples.tsv column names will be stripped of excess whitespace, similar to the config.
 - ATAC-seq pipeline removing the final bams, keeping the unsorted one
+>>>>>>> develop
```
Comment (Maarten-vd-Sande) [inline CHANGELOG.md:51]: :scream:

File: seq2science/rules/configuration_explain.smk [hunk 2]
```diff
@@ -176,4 +176,6 @@ else:
         "bed_slop": f"All summits were extended with {config.get('slop')} bp to get a consensus peakset.",
         "coverage_table": f"Finally, a count table from the consensus peakset with gimmemotifs.",  # already cited in "combine_peaks"
         "kb_seurat_pp": f"scRNA count post-processing was performed using the {hyperref('seurat')} based {hyperref('kb_seurat_pp')} markdown.",
+        "export_seurat_obj": f"scRNA count tables and meta-data were stored as {hyperref('seurat')} objects and stored in RDS format.",
+        "merge_seurat_obj": f"scRNA {hyperref('seurat')} objects were merged and stored in RDS format."
```
Comment (Maarten-vd-Sande) [inline seq2science/rules/configuration_explain.smk:180]: Not sure if this actually adds clarity. The idea is that the explanation that follows out of s2s is like a stub for the material and methods section. I don't think this is necessary

**Review by Maarten-vd-Sande — CHANGES_REQUESTED** [review]
Nice, two questions:

1) does it make sense for the kb_seurat_pp to use the merged object?
2) is it possible/easy to not use .rds format? .rds is R specific and thus not portable.. If it's not trivial to change, it's not worth the effort!

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +12,10 @@ All changed fall under either one of these types: `Added`, `Changed`, `Deprecate
 
 - added generic stats to the MultiQC report about the assembly, which might help pin point problems with the assembly used.
 - added the slop parameter to the config.yaml of atac-seq and chip-seq workflows, just so they are more visible.
+- added export_seurat_obj rule to read kb output to Seurat objects and store in RData format.
+- added merge_seurat_obj rule to merge single Seurat objects and store in RData format.
+- exporting/merging can be enabled by setting export_seu_objs: true
+- helper Seurat R scripts
```
[the remark this answers is not in the document]
Comment (tilschaef) [reply CHANGELOG.md:18]: 1. Technically yes, however, I am considering replacing kb_seurat_pp with a more accessible workflow in the future. 
2. I have to look into that but kb is able to provide the counts in adta format which gives the same sort of convenience for python

File: seq2science/rules/configuration_explain.smk (hunk 2 above)
Comment (tilschaef) [reply seq2science/rules/configuration_explain.smk]: Removed

File: seq2science/rules/quantification.smk [hunk 4]
```diff
@@ -321,8 +322,45 @@ elif config["quantifier"] == "kallistobus":
                   exit 1
                 fi  
                 """                
-                 
-                 
+                
+                
+    rule export_seurat_obj:
+        input:
```
[the remark this answers is not in the document]
Comment (tilschaef) [reply seq2science/rules/quantification.smk:331]: Added docstrings to rules

File: seq2science/scripts/seurat/read_kb_counts.R [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+} else if (isvelo) {
+  seu.sf <- CreateSeuratObject(counts = read_count_output(kb_dir, name="spliced"), assay = "sf", project = sample)
+  seu.uf <- CreateSeuratObject(counts = read_count_output(kb_dir, name="unspliced"), assay = "uf", project = sample)
+  seu.sf@meta.data <- prep_cell_meta(seu.sf, sample_sheet)
+  seu.uf@meta.data <- prep_cell_meta(seu.uf, sample_sheet)
+  seu_objs <- c(seu.sf, seu.uf)
+  names(seu_objs) <- c("sf","uf")
+  saveRDS(seu_objs, file = rds) 
+    
+} else {
```
Comment (siebrenf) [inline seq2science/scripts/seurat/read_kb_counts.R:80]: what is option 3? (just for clarity)

File: seq2science/scripts/seurat/merge_seurat_objs.R [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+# log all variables for debugging purposes
+cat('# variables used for this analysis:\n')
+cat('seu_dir      <- "', seu_dir,         '"\n', sep = "")
+cat('rds          <- "', rds,             '"\n', sep = "")
+cat('isvelo       <- "', isvelo,          '"\n', sep = "")
+
+cat('\n')
+
+# List RData files   
+print(seu_rdata)
```
Comment (siebrenf) [inline seq2science/scripts/seurat/merge_seurat_objs.R:26]: move to line 22 as 
```
cat('seu_rdata    <- c("', paste(seu_rdata, collapse='","'), '")\n', sep = "")
```

should look like this in the logs:
image

**Review by siebrenf — APPROVED** [review]
Great use of functions & comments! Haven't worked with Seurat yet, but I can follow along nicely.

my comments are meant as tips :)

File: seq2science/scripts/seurat/read_kb_counts.R (hunk 5 above)
Comment (tilschaef) [reply seq2science/scripts/seurat/read_kb_counts.R:80]: Option 3 is the default mode of kb,  scRNA quantification, when you do not provide the "--workflow" argument.

File: seq2science/scripts/seurat/merge_seurat_objs.R (hunk 6 above)
Comment (tilschaef) [reply seq2science/scripts/seurat/merge_seurat_objs.R]: Thanks for the tip 👍

**Review by Maarten-vd-Sande — APPROVED** [review]

**Pull request merged**